test/replication-luatest/election_strict_fencing_test.lua

```diff
@@ -0,0 +1,108 @@
+local luatest = require('luatest')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+local g = luatest.group('fencing_mode', {
+    {election_fencing_mode = 'soft'}, {election_fencing_mode = 'strict'}})
+
+local function promote(server)
+    luatest.helpers.retrying({}, function()
+        luatest.assert(server:exec(function()
+            pcall(box.ctl.promote)
+            if box.info.election.state ~= 'leader' then
+                return false
+            end
+            if box.info.synchro.queue.owner ~= box.info.id then
+                return false
+            end
+            return true
+        end))
+    end)
+    server:exec(function() box.ctl.wait_rw() end)
+end
+
+local function wait_sync(leader, servers)
+    local vclock = leader:get_vclock()
+    vclock[0] = nil
+    for _, server in ipairs(servers) do
+        server:wait_vclock(vclock)
+    end
+end
+
+local function box_cfg_update(servers, cfg)
+    for _, server in ipairs(servers) do
+        server:exec(function(cfg) box.cfg(cfg) end, {cfg})
+    end
+end
+
+g.before_all(function(g)
+    g.box_cfg = {
+        election_mode = 'manual',
+        election_timeout = 0.1,
+        replication = {
+            server.build_instance_uri('server_1'),
+            server.build_instance_uri('server_2'),
+        },
+        replication_synchro_quorum = 2,
+        replication_synchro_timeout = 0.1,
+        replication_timeout = 0.1,
+    }
+
+    g.cluster = cluster:new({})
+    g.server_1 = g.cluster:build_and_add_server(
+        {alias = 'server_1', box_cfg = g.box_cfg})
+
+    g.box_cfg.read_only = true
+    g.server_2 = g.cluster:build_and_add_server(
+        {alias = 'server_2', box_cfg = g.box_cfg})
+
+    g.cluster:start()
+    g.cluster:wait_fullmesh()
+    promote(g.server_1)
+    wait_sync(g.server_1, g.cluster.servers)
+end)
+
+g.after_all(function(g)
+    g.server_2:stop()
+    g.server_1:stop()
+end)
+
+g.test_fencing = function(g)
+    local timeout = 0.5
+    box_cfg_update(g.cluster.servers, {
+        election_fencing_mode = g.params.election_fencing_mode,
+        replication_timeout = timeout,
+    })
+    box_cfg_update({g.server_2}, {
+        election_mode = 'candidate',
+        replication_synchro_quorum = 1,
+    })
+    promote(g.server_1)
+    wait_sync(g.server_1, g.cluster.servers)
+
+    -- Wait for first heartbeat
+    require('fiber').sleep(timeout)
+
+    local old_term = g.server_1:election_term()
+
+    g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_COIO_WRITE_DROP', true)
```

Comment:
I saw the old conversation about why `ERRINJ_RELAY_SEND_DELAY` injection couldn't be used, but still we can't proceed with the injection like in the patch, because it simulates an impossible case. The alternatives I tried to propose somewhere above, will recite them here:
- Think about prolonging yields happening during replication reconfig;
- Add an injection to the second server's applier instead of the first server's relay;
- Add an injection to the relay which would break the connection.

Maybe there are more ways. But it is important that the final one should be in theory possible without an injection too.

Reply:
I want something like:
```bash
sudo iptables -I INPUT -p tcp --sport 3301 -j DROP
sudo iptables -I INPUT -p tcp --dport 3301 -j DROP
```
but for UNIX sockets, used in tests. I've decided to write a proxy, capable of providing such functionality for luatest. Will be back, when it works.